src/hotspot/share/opto/loopPredicate.cpp

```diff
@@ -797,7 +797,8 @@ BoolNode* PhaseIdealLoop::rc_predicate(IdealLoopTree* loop, Node* ctrl, int scal
   overflow = false;
   Node* max_idx_expr = nullptr;
   const TypeInt* idx_type = TypeInt::INT;
-  if ((stride > 0) == (scale > 0) == upper) {
+  // same signs and upper, or different signs and not upper.
+  if (((stride > 0) == (scale > 0)) == upper) {
```

Comment:
This is rather l33t code, but I guess it's OK with the comment. This
```suggestion
  _Bool same_signs = (stride > 0) == (scale > 0);
  if ((same_signs & upper)
      || (!same_signs && !upper)) {
```
generates slightly more code with GCC -O2.  I'd be happy with either.

Reply:
I agree it's a little odd, but I don't feel strongly about it, so leaving it as is.